File: 621-Task Scheduler.py
# formula method
# Runtime: 64 ms, faster than 79.65% of Python3 online submissions for Task Scheduler.
# Memory Usage: 13.3 MB, less than 5.03% of Python3 online submissions for Task Scheduler.
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        count = collections.Counter(tasks)
        most = count.most_common()[0][1]
        num_most = len([i for i, v in count.items() if v == most])
        time = (most - 1) * (n + 1) + num_most
        return max(time, len(tasks))


'''
（最多任务数-1）*（n + 1） + （相同最多任务的任务个数）
'''

File: Algorithms/C++/621-Task_Scheduler.cpp
// Runtime: 48 ms, faster than 98.45% of C++ online submissions for Task Scheduler.
// Memory Usage: 9.6 MB, less than 100.00% of C++ online submissions for Task Scheduler.
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        vector<int> temp(26, 0);
        for (char ch : tasks)
            ++temp[ch - 'A'];
        sort(temp.begin(), temp.end());
        int i = 25;
        while (i > 0 && temp[i] == temp[25])
            --i;
        return max((int) tasks.size(), (n + 1) * (temp[25] - 1) + 25 - i);
    }
};
